### ISSUE TITLE
BUG: Corner case dtype definition with zero-length strings fails.

### ISSUE THREAD

**Issue by mhvk** [numpy/numpy#27301]
### Describe the issue:

While trying to debug a corner case in which zero length strings are used (https://github.com/astropy/astropy/pull/16894), we realized that the following does not work generically for a given array `a` with non-zero length:
```
np.dtype((a.dtype, a.shape[1:]))
```

In particular, if one has an array that has only zero-length strings in it. This boils down to the following not working:
```
np.dtype(("S", ()))
```
Here, any shape tuple will fail, though one can pass in `0`.

### Reproduce the code example:

```python
import numpy as np
np.dtype(("S", ()))
```

### Error message:

```shell
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[5], line 1
----> 1 np.dtype(("S", ()))

ValueError: invalid itemsize in generic type tuple
```

### Python and NumPy Versions:

2.1.0
3.11.2 (main, May  2 2024, 11:59:08) [GCC 12.2.0]

**Comment by charris** [comment]
Is this a regression?

**Comment by mhvk** [comment]
> Is this a regression?

No, it is the same in numpy 1.24.

I think it goes back to when some new treatment for `S0` was instituted.

Note that there is another strange thing: if I pass through `np.array()` then if I copy the dtype gets changed too.  I'll add it on top.

**Comment by ngoldbaum** [comment]
I think it's always worked this way, at least back in NumPy 1.16 on Python 2.7 I get the same error anyway.

That said, it seems fair to let this case map to S0. Not that zero-sized dtypes make much sense, but I can see how one might fall out of code that is trying to generically memory-map string fields.

**Comment by mhvk** [comment]
Yes, I find it especially surprising that a copy changes the dtype. But not sure whether it is easy to solve...

**Comment by ngoldbaum** [comment]
Ah, missed the edit at the bottom. The copy changing the dtype is happening because of this code:

https://github.com/numpy/numpy/blob/37b3c3c2e0ebf4c018d92029bca0a0b2050634c1/numpy/_core/src/multiarray/ctors.c#L706-L718

I think maybe you found a corner case in the implementation of https://github.com/numpy/numpy/pull/23666, in the vein of https://github.com/numpy/numpy/issues/24043.

**Comment by mhvk** [comment]
Coming back to this, maybe time to ping @seberg -- I have no idea whether we can fiddle with that `_NPY_ARRAY_ALLOW_EMPTY_STRING` but really a copy should not change the dtype...

**Comment by seberg** [comment]
Not sure why that flag is related, that is array creation, not dtype creation.  You should be able to just remove the error in principle, although it might make sense to only allow it for `U0`, not `U`, since `U` should really be considered the DType and not the length 0 instance.

**Comment by mhvk** [comment]
> Not sure why that flag is related, that is array creation, not dtype creation. You should be able to just remove the error in principle, although it might make sense to only allow it for `U0`, not `U`, since `U` should really be considered the DType and not the length 0 instance.

I looked at the code and thought a bit more, and I think the basic problem is that right now there is no way to distinguish between "U" and "U0", since `PyDataType_ISUNSIZED` simply looks for `elsize == 0`. For legacy dtypes, there is no difference between the class and the instances, right? I guess in principle one could use a flag `NPY_UNSIZED` or so?

**Comment by seberg** [comment]
> For legacy dtypes, there is no difference between the class and the instances, right? 

There is, but it is a bit tricky.  We have a conversion function that can get DType or dtype that might help.  As of now, I am not sure it catches `U` vs. `U0` (it might, because what you get may be a different instance that is distinguishable).
So there may be a way to distinguish using `ISBUILTIN` or so, but would have to dig a bit to know.
